internal/controller/controller.go

```diff
@@ -0,0 +1,228 @@
+		ReportInterval:         intervals.ReportInterval(),
+		CacheSize:              traceHandlerCacheSize,
+		SamplesPerSecond:       c.config.SamplesPerSecond,
+		KernelVersion:          kernelVersion,
+		HostName:               hostname,
+		IPAddress:              sourceIP,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to start reporting: %w", err)
+	}
+	c.reporter = rep
+
+	metrics.SetReporter(rep)
+
+	// Now that set the initial host metadata, start a goroutine to keep sending updates regularly.
+	metadataCollector.StartMetadataCollection(ctx, rep)
+
+	// Load the eBPF code and map definitions
+	trc, err := tracer.NewTracer(ctx, &tracer.Config{
+		Reporter:               rep,
+		Intervals:              intervals,
+		IncludeTracers:         includeTracers,
+		FilterErrorFrames:      !c.config.SendErrorFrames,
+		SamplesPerSecond:       c.config.SamplesPerSecond,
+		MapScaleFactor:         int(c.config.MapScaleFactor),
+		KernelVersionCheck:     !c.config.NoKernelVersionCheck,
+		DebugTracer:            c.config.VerboseMode,
+		BPFVerifierLogLevel:    uint32(c.config.BpfVerifierLogLevel),
+		ProbabilisticInterval:  c.config.ProbabilisticInterval,
+		ProbabilisticThreshold: c.config.ProbabilisticThreshold,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to load eBPF tracer: %w", err)
+	}
+	c.tracer = trc
+	log.Printf("eBPF tracer loaded")
+
+	now := time.Now()
+	// Initial scan of /proc filesystem to list currently active PIDs and have them processed.
+	if err = trc.StartPIDEventProcessor(ctx); err != nil {
+		log.Errorf("failed to list processes from /proc: %v", err)
+	}
+	metrics.Add(metrics.IDProcPIDStartupMs, metrics.MetricValue(time.Since(now).Milliseconds()))
+	log.Debug("Completed initial PID listing")
+
+	// Attach our tracer to the perf event
+	if err := trc.AttachTracer(); err != nil {
+		return fmt.Errorf("failed to attach to perf event: %w", err)
+	}
+	log.Info("Attached tracer program")
+
+	if c.config.ProbabilisticThreshold < tracer.ProbabilisticThresholdMax {
+		trc.StartProbabilisticProfiling(ctx)
+		log.Printf("Enabled probabilistic profiling")
+	} else {
+		if err := trc.EnableProfiling(); err != nil {
+			return fmt.Errorf("failed to enable perf events: %w", err)
+		}
+	}
+
+	if err := trc.AttachSchedMonitor(); err != nil {
+		return fmt.Errorf("failed to attach scheduler monitor: %w", err)
+	}
+
+	// This log line is used in our system tests to verify if that the agent has started. So if you
+	// change this log line update also the system test.
+	log.Printf("Attached sched monitor")
+
+	if err := startTraceHandling(ctx, rep, intervals, trc, traceHandlerCacheSize); err != nil {
+		return fmt.Errorf("failed to start trace handling: %w", err)
+	}
+
+	return nil
+}
+
+// Shutdown stops the controller
+func (c *Controller) Shutdown() {
+	log.Info("Stop processing ...")
+	if c.tracer != nil {
+		c.tracer.Close()
+	}
+
+	if c.reporter != nil {
+		c.reporter.Stop()
+	}
```

Comment:
I don't know if this has unwanted side-effects (maybe @rockdaboot can clarify) but we're changing the order here compared to what it used to be (first we'd stop `reporter` in `main` and `tracer` through a `defer`). If we want to keep the original semantics, I'd use the original order.

```suggestion
	log.Info("Stop processing ...")	
	if c.reporter != nil {
		c.reporter.Stop()
	}
	
	if c.tracer != nil {
		c.tracer.Close()
	}

```